Buddies/Models/User.swift

```diff
@@ -0,0 +1,143 @@
+//
+//  User.swift
+//  Buddies
+//
+//  Created by Jake Thurman on 2/12/19.
+//  Copyright © 2019 Jack and the Beans. All rights reserved.
+//
+
+import Foundation
+import Firebase
+
+typealias UserId = String
+
+protocol UserInvalidationDelegate {
+    func onInvalidateUser(user: User)
+}
+
+class User {
+    let delegate: UserInvalidationDelegate?
+    
+    // MARK: Immutable Properties
+    let uid : UserId
+    let email : String?
+    let isAdmin : Bool
+    let facebookId : String?
+    let dateJoined : Date
+    let notificationToken : String?
+    
+    // MARK: Mutable Properties
+    var imageUrl : String { didSet { onChange("image_url", imageUrl) } }
+    var name : String { didSet { onChange("name", name) } }
+    var bio : String { didSet { onChange("bio", bio) } }
+    var favoriteTopics : [String] { didSet { onChange("favorite_topics", favoriteTopics) } }
+    var location : GeoPoint? { didSet { onChange("favorite_topics", location) } }
+    
+    // MARK: User Settings
+    var shouldSendJoinedActivityNotification : Bool {
+        didSet {
+            onChange("should_send_joined_activity_notification",
+                     shouldSendJoinedActivityNotification)
+        }
+    }
+    var shouldSendActivitySuggestionNotification : Bool {
+        didSet {
+            onChange("should_send_activity_suggestion_notification",
+                     shouldSendActivitySuggestionNotification)
+        }
+    }
+    
+    // MARK: Blocking
+    var blockedUsers : [UserId] { didSet { onChange("blocked_users", blockedUsers) } }
+    var blockedActivities : [ActivityId] { didSet { onChange("blocked_activities", blockedActivities) } }
+    let blockedBy : [UserId] // Shouldn't be updated directly! Automatic inverse of blocked_users.
```

Comment:
Should this property be `readonly` then?

Reply:
it's a let idiot